### PR TITLE
Changed the field `name` to `label`

### PR DESCRIPTION
**Pull request by xgreenx**
Updated the code generation of ABI. Instead of "name" in messages, constructors, arguments, events we are using "label".
Changed the behavior of "label" for messages and constructors, now it is a string instead of an array. 
For trait's messages, we will use the "TRAIT_IDENT::MESSAGE_IDENT" format of representation.

This PR is related to https://github.com/paritytech/ink/issues/900.

### CODE REVIEW

**Comment by xgreenx** [comment]
@cmichi What do you think about updating `displayName` to be a string instead of an array too?

**Comment by cmichi** [comment]
Hey nice, that you created a PR for this!

> @cmichi What do you think about updating `displayName` to be a string instead of an array too?

:-) I actually had intended to create a ticket for this today! I agree that it's a good idea, but let's keep this to a separate issue. I'll create one where we track this.

File: README.md
Comment (cmichi) [inline README.md:224]:
```suggestion
| `#[ink(label = "..")]` | Applicable to ink! messages, ink! constructors and ink! implementation blocks. | Specifies a concrete label of a method/trait inside the metadata. This allows a contract author to precisely control the labeling of methods inside of ABI making it possible to rename their methods without breakage. |
```

File: crates/lang/ir/src/ir/attrs.rs
Comment (cmichi) [inline crates/lang/ir/src/ir/attrs.rs:390]:
```suggestion
    /// Applied to ink! constructors, messages or implementation blocks
    /// with traits to manually control their labeling inside the metadata.
```

File: crates/lang/ir/src/ir/item_impl/message.rs
Comment (cmichi) [inline crates/lang/ir/src/ir/item_impl/message.rs:107]:
```suggestion
    /// An optional user provided label of the method for the metadata.
```

File: crates/lang/ir/src/ir/item_impl/message.rs
Comment (cmichi) [inline crates/lang/ir/src/ir/item_impl/message.rs:111]:
```suggestion
    /// This overwrites the default label of the method inside the metadata.
```

File: crates/lang/ir/src/ir/item_impl/mod.rs
Comment (cmichi) [inline crates/lang/ir/src/ir/item_impl/mod.rs:83]:
```suggestion
    /// A label of the trait used in the metadata.
```

File: crates/lang/ir/src/ir/item_impl/mod.rs
Comment (cmichi) [inline crates/lang/ir/src/ir/item_impl/mod.rs:88]:
```suggestion
    /// The user can provide a label of the trait which will be used inside the metadata.
    /// There is no requirement to implement this trait though.
```

File: crates/lang/ir/src/ir/item_impl/mod.rs
Comment (cmichi) [inline crates/lang/ir/src/ir/item_impl/mod.rs:373]:
```suggestion
    /// If the label is `""`, it means that the user wants to remove the trait labeling in the metadata.
```

File: crates/lang/macro/src/lib.rs
Comment (cmichi) [inline crates/lang/macro/src/lib.rs:383]:
```suggestion
///     is used during generation of the metadata (ABI).
```

File: crates/lang/macro/src/lib.rs
Comment (cmichi) [inline crates/lang/macro/src/lib.rs:385]:
```suggestion
///     An ink! smart contract author can control the labeling in the metadata using the `label` flag.
///     An example is shown below:
```

File: crates/lang/macro/src/lib.rs
Comment (cmichi) [inline crates/lang/macro/src/lib.rs:406]:
```suggestion
///         # #[ink(label = "my_flip")] // You can also specify the label out-of-line.
```

File: crates/lang/macro/tests/ui/pass/12-attributes.rs
Comment (cmichi) [inline crates/lang/macro/tests/ui/pass/12-attributes.rs:7]:
```suggestion
        // Only `#[ink(message)]` and `#[ink(constructor)]` are allowed in trait definitions
```

File: crates/metadata/src/specs.rs
Comment (cmichi) [inline crates/metadata/src/specs.rs:225]:
```suggestion
    /// In case of a trait provided constructor the label is prefixed with the trait label.
```

File: crates/metadata/src/specs.rs
Comment (cmichi) [inline crates/metadata/src/specs.rs:258]:
```suggestion
    /// In case of a trait provided constructor the label is prefixed with the trait label.
```

**Review by cmichi — CHANGES_REQUESTED** [review]
Looks good on first glance ‒ great that you implemented this!

Got some language remarks. I guess the PSP-22 should also be adapted at some point in the future.

File: README.md
Comment (Robbepop) [reply README.md:224]: > | `#[ink(label = "..")]` | Applicable to ink! messages, ink! constructors and ink! implementation blocks. 

Why do we have the need to label ink! implementation blocks? What is the effect of it?

Comment (xgreenx) [reply README.md]: Marking the implementation block allows you to specify the prefix before the message or constructor. 
The user can use it to remove the prefix of trait, or change the prefix of trait, or add a prefix for the impl section without the trait.

In this [PR](https://github.com/paritytech/ink/pull/859) I described the case when we are using "change the prefix of trait" behavior.

**Comment by Robbepop** [comment]
@xgreenx please rebase this on the current `master` branch so we might be ready for another round of review and merge it.

File: README.md
Comment (Robbepop) [reply README.md]: I am not sure I like this feature for trait implementation blocks since I see a potential for a lot of confusion. What is an actual concrete use case there?

File: crates/lang/ir/src/ir/attrs.rs
Comment (Robbepop) [inline crates/lang/ir/src/ir/attrs.rs:793]: This needs a check to be a valid Rust identifier.
You can check this by using
```rust
syn::parse_str::<syn::Ident>()
    .unwrap_or_else(|error| {
        panic!("encountered invalid non-identifier namespace argument: {}", error)
    })
```

Comment (Robbepop) [reply crates/lang/ir/src/ir/attrs.rs:793]: Actually instead of panicking you should return a proper `syn::Error` via `format_err!` instead.

Comment (Robbepop) [reply crates/lang/ir/src/ir/attrs.rs:793]: https://github.com/paritytech/ink/pull/931/files#diff-024136e95afaa59cacb0e87a4287e1e425de563db108032ba4b2d40df2c160eaR770

File: README.md
Comment (xgreenx) [reply README.md]: The actual use case is:
We have a trait with the name `Erc20External`. During generation of labels for methods ink! adds `Erc20External::` prefix. With this prefix it generates `Erc20External::transfer` label. The idea is to label the impl section with the `Erc20` label to generate the `Erc20::transfer` for the method.

We can remove the `label` attribute for the impl section if the `label` attribute also overrides the prefix(in the current implementation it only overrides the label of method and trait prefix affects it).

Comment (cmichi) [reply README.md]: > I am not sure I like this feature for trait implementation blocks since I see a potential for a lot of confusion. What is an actual concrete use case there?

@Robbepop Can you elaborate on where you see potential for confusion here? If you haven't seen yet, the [issue for this PR](https://github.com/paritytech/ink/issues/900) states some more of the motivation:

>For traits, ink! generates a field in the form of `"name": ["BaseErc20", "new"]`. The `BaseErc20` hereby is a trait implemented by the contract. This exposes internal implementation details of the contract without the developer having any influence over it.

File: crates/lang/ir/src/ir/attrs.rs
Comment (xgreenx) [reply crates/lang/ir/src/ir/attrs.rs:793]: @Robbepop  I'm waiting for your decision in the [thread](https://github.com/paritytech/ink/pull/923#discussion_r714503215).
Because it will affect this part too=)

Comment (Robbepop) [reply crates/lang/ir/src/ir/attrs.rs:793]: I cannot tell either what the better design is.
Let us summarize what you propose:

1) `#[ink(label = "foo")]` on an ink! implementation block is a prefix for `#[ink(label = "bar")]` on ink! messages in that same block. So both combined you end up with a label `foo::bar`.
2) Any `#[ink(label = "bar")]` on an ink! message or constructor in an ink! implementation block will overwrite the label of the ink! message or constructor entirely. Any `#[ink(label = "foo")]` on the ink! implementation block will be silently ignored.

Comment (xgreenx) [reply crates/lang/ir/src/ir/attrs.rs:793]: 1. Yes, the label on ink! the message will overwrite only suffix(doesn't affect the prefix from the label on ink! impl block).
2. Yes, the label on ink! the message will overwrite entirely(affect the prefix from the label on ink! impl block).
3. `#[ink(label = "foo")]` on implementation block will be assumed as an error(it means remove support of the label for impl block).  `#[ink(label = "bar")]` on an ink! message in ink! implementation block will overwrite the default behavior(when the label of ink! message is a combination of trait name and message label).

I implemented the first case.
For me, the second case seems similar to the behavior of the `selector` attribute(if you defined the `selector`, it will ignore `namespace`.
The third case I created after your idea to remove the label for ink! implementation block=)

The idea of the `label` attribute is to add the ability to change the generated ABI. 
For example, in the code, I have two traits: One for internal usage(without ink! messages); another with ink! messages. Both traits have the method with the same name. It requires the developer in the code to explicitly specify the trait which he wants to use. To avoid this, I want to change the name of the method in the trait with ink! messages(I can't change the name in internal trait). But I will change the label of the method in ABI. In this case, the `label` attribute will help me to generate the right ABI.

**Comment by cmichi** [comment]
I've lost a bit track of where we are with this PR and where the discussion went.

@xgreenx Could you maybe quickly summarize what decision needs to be made for you to be able to finish the PR?

**Comment by xgreenx** [comment]
We need to decide do we need the `label` attribute or not. You think that [better to add](https://github.com/paritytech/ink/issues/900#issuecomment-949449714), @Robbepop thinks it is[ better to not add](https://github.com/paritytech/ink/issues/900#issuecomment-949433312) additional attributes=) [My thoughts](https://github.com/paritytech/ink/issues/900#issuecomment-949518025)

**Comment by Robbepop** [comment]
> We need to decide do we need the `label` attribute or not. You think that [better to add](https://github.com/paritytech/ink/issues/900#issuecomment-949449714), @Robbepop thinks it is[ better to not add](https://github.com/paritytech/ink/issues/900#issuecomment-949433312) additional attributes=) [My thoughts](https://github.com/paritytech/ink/issues/900#issuecomment-949518025)

I am strictly against adding the label attribute. the metadata changes are ok. i see no reason in adding a feature where there is no strict requirement and where we are not 100% sure about its design.

**Comment by cmichi** [comment]
>i see no reason in adding a feature where there is no strict requirement and where we are not 100% sure about its design.

@Robbepop Could you please reply to the points brought up in the discussion [here](https://github.com/paritytech/ink/issues/900#issuecomment-949449714)?

**Comment by cmichi** [comment]
@xgreenx I've discussed it with @Robbepop and the outcome is that we'll only include the metadata changes for now and revisit the `#[ink(label)]` discussion at a later point. Would you be so kind to adapt your PR?

**Comment by xgreenx** [comment]
Sure, will adapt it for the master branch and will remove `label` related code=)

**Comment by xgreenx** [comment]
I've updated the PR. In ABI we still use "name" for storage layout and for types. Do I need to update it for storage layout? (because in the case of types we need to update `scale-info`)

File: crates/lang/codegen/src/generator/metadata.rs
Comment (xgreenx) [inline crates/lang/codegen/src/generator/metadata.rs:278]: Do you prefer this logic here or in `from_trait_and_label` on "spec" level?

What do you think about the idea that the namespace will affect the label too(not only selector)? If you agree then I will implement it like this https://github.com/paritytech/ink/blob/master/crates/lang/ir/src/ir/selector.rs#L108

**Comment by cmichi** [comment]
@xgreenx Could you merge `master` into this PR?

**Comment by cmichi** [comment]
>In ABI we still use "name" for storage layout and for types. Do I need to update it for storage layout? (because in the case of types we need to update scale-info)

Nope, no need to adapt those.

File: crates/lang/codegen/src/generator/metadata.rs
Comment (xgreenx) [reply crates/lang/codegen/src/generator/metadata.rs:278]: I meant, that I removed `from_trait_and_label ` from `spec` and implement the logic of concatenating(joining of trait name and method name by "::") here, on `generator` level.

Comment (cmichi) [reply crates/lang/codegen/src/generator/metadata.rs:278]: Alright, got it. It's fine here.

Do you want to implement the namespace affecting the label as well? Could also be done as a follow-up.

Comment (cmichi) [reply crates/lang/codegen/src/generator/metadata.rs:278]: I'll merge this now, I think the PR is fine in its current state and we've been putting you on hold for way too long :-).

Feel free to create a follow-up for including the namespace into the label.

**Comment by cmichi** [comment]
@xgreenx The `ink-waterfall` CI stage is expected to fail, since it builds our example contracts and tries to upload them via polkadot-js. Since that UI doesn't understand the new metadata format yet the upload fails. I've created an issue for it and will merge the PR after that one is solved. Feel free to implement the namespace inclusion in the meantime or as a follow-up, whatever works best for you :-).

File: crates/lang/codegen/src/generator/metadata.rs
Comment (xgreenx) [reply crates/lang/codegen/src/generator/metadata.rs:278]: I can implement that in follow-up PR=) @Robbepop What do you think about the namespace affecting the label?

**Review by cmichi — APPROVED** [review]

**Pull request merged**
